#!/usr/bin/env python
PACKAGE = "pcl_recognizer"

from dynamic_reconfigure.parameter_generator_catkin import *

gen = ParameterGenerator()
gen.add("str_param",    str_t,       0, "A string parameter",  "Hello World")
gen.add("recalculate", bool_t,      0, "Temporary trigger",  True)
stopat_enum = gen.enum([ gen.const("Load",      int_t, 0, "Load and view only"),
                       gen.const("Denoise",     int_t, 1, "Play with denoising"),
                       gen.const("Normals",     int_t, 2, "Play with normals"),
                       gen.const("Keypoints",   int_t, 3, "Play with keypoints"),
                       gen.const("Descriptors", int_t, 4, "Play with descriptors"),
                       gen.const("Grouping",    int_t, 5, "Play with grouping"),
                       gen.const("All", int_t, 6, "Play with all")],
                     "Enum for selecting recalculation range")
gen.add("stop_at", int_t, 0, "Recalculate until selected step", 2, 0, 6, edit_method=stopat_enum)
gen.generate(PACKAGE, "pcl_recognizer", "Recognizer")

key_gen = ParameterGenerator()
method_enum = key_gen.enum([ key_gen.const("UniformSampling",      int_t, 0, "Uniform sampling"),
                         key_gen.const("ISS3D",     int_t, 1, "ISS3D")],
                         "An enum to set keypoint method" )
key_gen.add("method", int_t, 0, "Keypoint method selection", 1, 0, 1, edit_method=method_enum)
key_gen.add("uniform_radius",     double_t,    0, "Model uniform sampling radius (default 0.01)",  .01,  0, .5)
key_gen.add("iss_non_max_radius", double_t,    0, "Model uniform sampling radius (default 0.012)", .012, 0, .05)
key_gen.add("iss_border_radius",  double_t,    0, "Model uniform sampling radius (default 0.004)", .004, 0, .05)
key_gen.add("iss_salient_radius", double_t,    0, "Model uniform sampling radius (default 0.015)", .015, 0, .05)
key_gen.generate(PACKAGE, "pcl_recognizer", "Keypoint")

desc_gen = ParameterGenerator()
normal_method_enum = desc_gen.enum([ desc_gen.const("OMP",      int_t, 0, "OMP"),
                                desc_gen.const("INT",     int_t, 1, "Integral images"),
                                desc_gen.const("MLS",     int_t, 2, "MLS")],
                                 "An enum to set normals method" )
desc_gen.add("normal_method",   int_t,    0, "Normal method selection", 0, 0, 2, edit_method=normal_method_enum)
desc_gen.add("normal_ksize",    int_t,    0, "Normal k neighbourhood (default 0)", 0, 0, 100)
desc_gen.add("normal_rad",    double_t,   0, "Normal radius (default 0.03)", 0.03, 0, .5)
desc_gen.add("normal_omp_threads",    double_t,   0, "Thread pool size (default 10)", 10, 0, 1000)
desc_gen.add("normal_int_smoothing", double_t,   0, "Integral image smoothing (default 10.0)", 10.0, 0, 100.0)
desc_gen.add("normal_int_maxdepth", double_t,   0, "Max depth change factor (default 0.02)", 0.02, 0, .5)
desc_gen.add("descr_rad",    double_t,    0, "Descriptor radius (default 0.02)", .02, 0, .5)
desc_gen.add("rf_rad",       double_t,    0, "Reference frame radius (default 0.015)", .015, 0, .5)
desc_gen.generate(PACKAGE, "pcl_recognizer", "Descriptor")

match_gen = ParameterGenerator()
match_gen.add("cg_size",      double_t,    0, "Cluster size (default 0.01)", .01, 0, .5)
match_gen.add("cg_thresh",    double_t,    0, "Clustering threshold (default 5)", 5, 0, 100)
match_gen.add("use_hough",    bool_t,      0, "Grouping method",  True)
match_gen.generate(PACKAGE, "pcl_recognizer", "Grouping")

view_gen = ParameterGenerator()
view_gen.add("input",   bool_t,   0, "View input cloud",  True)
view_gen.add("keypoints",   bool_t,   0, "View keypoints",  True)
view_gen.add("normals",   bool_t,   0, "View normals",  False)
view_gen.add("descriptors",   bool_t,   0, "View descriptors",  True)
view_gen.add("update",   bool_t,   0, "Temporary trigger",  True)

exit(view_gen.generate(PACKAGE, "pcl_recognizer", "Viewer"))


